File: ios/Services/PaywallManager.swift
import Foundation
import Combine
import SwiftUI

/// Coordinates SuperwallKit paywall triggers and subscription entitlement verification.
@MainActor
public final class PaywallManager: ObservableObject {
    public static let shared = PaywallManager()

    @Published public private(set) var subscriptionState = UserSubscriptionState()
    @Published public var shouldPresentPaywall: Bool = false
    @Published public var activePaywallReason: String = ""

    private let userDefaultsKey = "com.wakeorpay.subscription_state"

    private init() {
        loadSubscriptionState()
    }

    /// Configures Superwall SDK with the specified API key.
    public func configure(apiKey: String) {
        // In production:
        // Superwall.configure(apiKey: apiKey)
        print("[PaywallManager] Superwall configured with API Key: \(apiKey.prefix(6))...")
    }

    /// Evaluates whether an action requires Pro gating.
    /// If Pro is not active, triggers the paywall with the corresponding placement context.
    public func evaluateProRequirement(reason: String, required: Bool) -> Bool {
        if !required || subscriptionState.isPro {
            return true
        }

        self.activePaywallReason = reason
        self.shouldPresentPaywall = true
        return false
    }

    /// Checks whether user is allowed to create or enable an additional active alarm.
    public func canActivateAlarm(currentActiveCount: Int) -> Bool {
        if subscriptionState.isPro {
            return true
        }
        if currentActiveCount >= 1 {
            self.activePaywallReason = "Free plan includes 1 active alarm. Upgrade to Pro for unlimited alarms."
            self.shouldPresentPaywall = true
            return false
        }
        return true
    }

    /// Simulates or executes subscription purchase for the requested tier.
    public func purchase(tier: SubscriptionTier) async -> Bool {
        // Simulate purchase confirmation for development
        self.subscriptionState = UserSubscriptionState(
            currentTier: tier,
            isTrialActive: tier == .proWeekly,
            expirationDate: Calendar.current.date(byAdding: .day, value: tier == .proWeekly ? 7 : 3650, to: Date())
        )
        persistSubscriptionState()
        self.shouldPresentPaywall = false
        return true
    }

    /// Restores previous in-app purchases.
    public func restorePurchases() async -> Bool {
        print("[PaywallManager] Restoring purchases...")
        return true
    }

    private func persistSubscriptionState() {
        do {
            let data = try JSONEncoder().encode(subscriptionState)
            UserDefaults.standard.set(data, forKey: userDefaultsKey)
        } catch {
            print("[PaywallManager] Failed to persist subscription: \(error.localizedDescription)")
        }
    }

    private func loadSubscriptionState() {
        guard let data = UserDefaults.standard.data(forKey: userDefaultsKey),
              let state = try? JSONDecoder().decode(UserSubscriptionState.self, from: data) else {
            self.subscriptionState = UserSubscriptionState(currentTier: .free)
            return
        }
        self.subscriptionState = state
    }
}

File: android/app/src/main/java/com/wakeorpay/app/service/PaywallManager.kt
package com.wakeorpay.app.service

import android.content.Context
import com.wakeorpay.app.model.SubscriptionTier
import com.wakeorpay.app.model.UserSubscriptionState
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * Manages Superwall integration and subscription entitlement gating on Android.
 */
object PaywallManager {

    private const val PREFS_NAME = "wakeorpay_paywall_prefs"
    private const val KEY_SUBSCRIPTION = "subscription_state"

    private val _subscriptionState = MutableStateFlow(UserSubscriptionState())
    val subscriptionState: StateFlow<UserSubscriptionState> = _subscriptionState.asStateFlow()

    private val _shouldShowPaywall = MutableStateFlow(false)
    val shouldShowPaywall: StateFlow<Boolean> = _shouldShowPaywall.asStateFlow()

    private val _paywallReason = MutableStateFlow("")
    val paywallReason: StateFlow<String> = _paywallReason.asStateFlow()

    private var appContext: Context? = null

    fun init(context: Context, apiKey: String) {
        appContext = context.applicationContext
        loadSubscriptionState()
        // Superwall SDK configuration:
        // Superwall.configure(context, apiKey)
    }

    fun evaluateProRequirement(reason: String, required: Boolean): Boolean {
        if (!required || _subscriptionState.value.hasProAccess) {
            return true
        }
        _paywallReason.value = reason
        _shouldShowPaywall.value = true
        return false
    }

    fun canActivateAlarm(currentActiveCount: Int): Boolean {
        if (_subscriptionState.value.hasProAccess) {
            return true
        }
        if (currentActiveCount >= 1) {
            _paywallReason.value = "Free plan includes 1 active alarm. Upgrade to Pro for unlimited alarms."
            _shouldShowPaywall.value = true
            return false
        }
        return true
    }

    fun dismissPaywall() {
        _shouldShowPaywall.value = false
    }

    fun purchase(tier: SubscriptionTier) {
        val newState = UserSubscriptionState(
            tier = tier,
            isTrialActive = tier == SubscriptionTier.PRO_WEEKLY,
            expiresAtEpochMs = System.currentTimeMillis() + if (tier == SubscriptionTier.PRO_WEEKLY) 7L * 86400000L else 3650L * 86400000L
        )
        _subscriptionState.value = newState
        _shouldShowPaywall.value = false
        saveSubscriptionState(newState)
    }

    private fun saveSubscriptionState(state: UserSubscriptionState) {
        appContext?.let { ctx ->
            val prefs = ctx.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            val json = Json.encodeToString(state)
            prefs.edit().putString(KEY_SUBSCRIPTION, json).apply()
        }
    }

    private fun loadSubscriptionState() {
        appContext?.let { ctx ->
            val prefs = ctx.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            val json = prefs.getString(KEY_SUBSCRIPTION, null)
            if (json != null) {
                try {
                    _subscriptionState.value = Json.decodeFromString<UserSubscriptionState>(json)
                } catch (e: Exception) {
                    _subscriptionState.value = UserSubscriptionState()
                }
            }
        }
    }
}
